drivers/photo-electronic/adpd188/adpd188.c

```diff
@@ -0,0 +1,871 @@
+ * @param reg_addr - The register address.
+ * @param reg_val - The new value of the register.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t adpd188_reg_write(struct adpd188_dev *dev, uint8_t reg_addr,
+			  uint16_t reg_val)
+{
+	int32_t ret;
+	uint8_t buff[] = {0, 0, 0};
+
+	buff[1] = (reg_val & 0xFF00) >> 8;
+	buff[2] = reg_val & 0x00FF;
+
+	if(dev->phy_opt == ADPD188_SPI) {
+		buff[0] = (reg_addr << 1) | 1;
+		ret = spi_write_and_read(dev->phy_desc, buff, 3);
+	} else if(dev->phy_opt == ADPD188_I2C) {
+		buff[0] = reg_addr;
+		ret = i2c_write(dev->phy_desc, buff, 3, 1);
+	} else {
+		ret = FAILURE;
+	}
+	if(ret != SUCCESS)
+		return FAILURE;
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Get the mode of operation of the ADPD188.
+ * @param dev - The ADPD188 descriptor.
+ * @param mode - The cur mode of operation. Values can be:
+ *                   - ADPD188_STANDBY;
+ *                   - ADPD188_PROGRAM;
+ *                   - ADPD188_NORMAL.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t adpd188_mode_get(struct adpd188_dev *dev, enum adpd188_mode *mode)
+{
+	int32_t ret;
+	uint16_t data;
+
+	ret = adpd188_reg_read(dev, ADPD188_REG_MODE, &data);
+	if(ret != SUCCESS)
+		return FAILURE;
+
+	*mode = data & ADPD188_MODE_MODE_MASK;
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Set the mode of operation of the ADPD188.
+ * @param dev - The ADPD188 descriptor.
+ * @param new_mode - The new mode of operation. Values can be:
+ *                   - ADPD188_STANDBY;
+ *                   - ADPD188_PROGRAM;
+ *                   - ADPD188_NORMAL.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t adpd188_mode_set(struct adpd188_dev *dev, enum adpd188_mode new_mode)
+{
+	uint16_t data;
+
+	data = new_mode & ADPD188_MODE_MODE_MASK;
+
+	return adpd188_reg_write(dev, ADPD188_REG_MODE, data);
+}
+
+/**
+ * @brief Get the number of bytes currently present in FIFO.
+ * @param dev - The ADPD188 descriptor.
+ * @param bytes_no - Number of bytes in the FIFO.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t adpd188_fifo_status_get(struct adpd188_dev *dev, uint8_t *bytes_no)
+{
+	int32_t ret;
+	uint16_t reg_data;
+
+	ret = adpd188_reg_read(dev, ADPD188_REG_STATUS, &reg_data);
+	if(ret != SUCCESS)
+		return FAILURE;
+
+	*bytes_no = (reg_data & ADPD188_STATUS_FIFO_SAMPLES_MASK) >>
+		    ADPD188_STATUS_FIFO_SAMPLES_POS;
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Empty the FIFO.
+ * @param dev - The ADPD188 descriptor.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t adpd188_fifo_clear(struct adpd188_dev *dev)
+{
+	int32_t ret;
+	uint16_t reg_data;
+
+	ret = adpd188_reg_read(dev, ADPD188_REG_STATUS, &reg_data);
+	if(ret != SUCCESS)
+		return FAILURE;
+	reg_data |= 0x8000;
+	/* Write 0 to the interrupt flags to not clear them unintentionally. */
+	reg_data &=
+		~(ADPD188_STATUS_SLOTA_INT_MASK | ADPD188_STATUS_SLOTB_INT_MASK);
```

Comment:
This looks a bit weird, maybe force an indentation where only `ADPD188_STATUS_SLOTB_INT_MASK);` is on the next line

Reply:
OK, can do that. I was thinking that not separating the mask components would make it easier to read.